node/silkworm/common/log.hpp

```diff
@@ -17,60 +17,79 @@
-
-// available verbosity levels
-enum class LogLevel { Trace, Debug, Info, Warn, Error, Critical, None };
-
-// change the logging output streams - default is (cerr, null_stream())
-//
-#define SILKWORM_LOG_STREAMS(stream1_, stream2_) silkworm::log_set_streams_((stream1_), (stream2_));
-
-// silence
-std::ostream& null_stream();
-
-//
-// Below are for access via macros ONLY.
-// Placing them in detail namespace prevents use of macros in nested namespaces of silkworm :(
-//
-extern LogLevel log_verbosity_;
-extern bool log_thread_enabled_;
-void log_set_streams_(std::ostream& o1, std::ostream& o2);
-class log_ {
+#include <filesystem>
+#include <sstream>
+
+#include <silkworm/common/terminal.hpp>
+
+namespace silkworm::log {
+
+//! \brief Available verbosity levels
+enum class Level {
+    None,      // Simple logging line with no severity (eg. build info)
+    Critical,  // An error there's no way we can recover from
+    Error,     // We encountered an error which we might be able to recover from
+    Warn,      // Something happened and user might have the possibility to amend the situation
+    Info,      // Info messages on regular operations
+    Debug,     // Debug information
+    Trace      // Trace calls to functions
+};
+
+//! \brief Holds logging configuration
+struct Settings {
+    bool log_std_out{false};    // Whether console logging goes to std::cout or std::cerr (default)
+    bool log_utc{false};        // Whether timestamps should be in UTC or imbue local timezone
+    bool log_nocolor{false};    // Whether to disable colorized output
+    bool log_threads{false};    // Whether to print thread ids in log lines
+    unsigned log_verbosity{4};  // Log verbosity level
+    std::string log_file;       // Log to file
+};
+
+//! \brief Initializes logging facilities
+void init(Settings& settings);
+
+//! \brief Sets logging verbosity
+void set_verbosity(Level level);
+
+//! \brief Sets a file output for log teeing
+void tee_file(std::filesystem::path path);
+
+class BufferBase {
   public:
-    explicit log_(LogLevel level) : level_(level) { log_mtx_.lock(); }
-    ~log_() { log_mtx_.unlock(); }
-    std::ostream& header_(LogLevel);
+    explicit BufferBase(Level level);
+    ~BufferBase() { flush(); }
+
+    // Accumulators
+    template <class T>
+    inline void append(T const& t) {
+        ss_ << t;
+    }
     template <class T>
-    std::ostream& operator<<(const T& message) {
-        return header_(level_) << message;
+    BufferBase& operator<<(T const& t) {
+        append(t);
+        return *this;
     }
 
-  private:
-    LogLevel level_;
-    static std::mutex log_mtx_;
+  protected:
+    void flush();
+    Level level_;
+    std::stringstream ss_;
 };
 
-}  // namespace silkworm
+template <Level level>
+class LogBuffer : public BufferBase {
+  public:
+    LogBuffer() : BufferBase(level){};
+};
+
+using TraceChannel = LogBuffer<Level::Trace>;
+using DebugChannel = LogBuffer<Level::Debug>;
```

Comment:
I would even drop "Channel" to simplify call sites to something like
```
log::Info() << "Initializing Light Cache for DAG epoch " << epoch_num;
```
instead of 
```
log::InfoChannel() << "Initializing Light Cache for DAG epoch " << epoch_num;
```

Reply:
I thought about that too but it causes an odd behavior on CLion.
Basically it (CLion) confuses the class with the value of LogLevel enum and improperly colorizes the line to "used" when it "might" be unused

Reply:
Not a problem if we prefix the enum values with `k` per Google's style guide.

Reply:
Done in 4a51321 and 4b131a1